### PR TITLE
[clang] Deduplicate the logic that only warns once when stack is almost full

### PR DESCRIPTION
**Pull request by bricknerb**
Zero diff in behavior.

### CODE REVIEW

**Comment by bricknerb** [comment]
This pull request is a replacement for https://github.com/llvm/llvm-project/pull/112371.

File: clang/include/clang/Basic/SingleWarningStackAwareExecutor.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_BASIC_SINGLE_WARNING_STACK_AWARE_EXECUTOR_H
+#define LLVM_CLANG_BASIC_SINGLE_WARNING_STACK_AWARE_EXECUTOR_H
+
+#include "clang/Basic/Diagnostic.h"
+
+namespace clang {
+class SingleWarningStackAwareExecutor {
```
Comment (cor3ntin) [inline clang/include/clang/Basic/SingleWarningStackAwareExecutor.h:21]: The primary functionality is to grow the stack, The name should be more reflective of that
Maybe `StackExhaustionHandler" or some such

**Review by ilya-biryukov — APPROVED** [review]
LGTM, but please fix the comment from @cor3ntin before submitting about renaming the class

**Review by ChuanqiXu9 — APPROVED** [review]

File: clang/include/clang/Basic/SingleWarningStackAwareExecutor.h (hunk 1 above)
Comment (bricknerb) [reply clang/include/clang/Basic/SingleWarningStackAwareExecutor.h]: Done.

**Review by cor3ntin — APPROVED** [review]
LGTM

**Pull request merged**